<a href="https://colab.research.google.com/github/KalinaMarkova/deep_learning_course_project/blob/main/04_Model_Training_Experiment_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fine-Grained Analysis of Propaganda in News Articles
## Notebook 04: Hierarchical Two-Stage Pipeline: Binary Span Detector & Technique Classifier - Experiment 2

In this notebook we will implement a Hierarchical Two-Model Architecture to detect propaganda in news articles. Instead of forcing a single model to solve boundary extraction and 14-class technique classification simultaneously, we divide the task into two specialized stages:

**Model A (Binary Span Detector)**: A token-classification model trained strictly on 3 labels (O, B-PROPAGANDA, I-PROPAGANDA). Its sole purpose is to draw precise bounding boxes around manipulative language spans.

**Model B (Technique Classifier)**: A sequence-classification model that takes the isolated text spans identified by Model A and categorizes them into their specific propaganda classes (e.g., Loaded Language, Slogans, Name Calling).

By separating the tasks, each model gets to specialize, which we expect could improve the performance of the final pipeline.

In [ ]:
!pip install -q transformers datasets seqeval evaluate

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 2.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 6.5 MB/s eta 0:00:00


In [33]:
from google.colab import drive
from datasets import load_from_disk
from transformers import AutoTokenizer, AutoModelForTokenClassification, TrainingArguments, Trainer, DataCollatorForTokenClassification, DataCollatorWithPadding, AutoModelForSequenceClassification
import evaluate
import numpy as np
import torch
import torch.nn as nn
import pandas as pd
import glob
import os
from datasets import Dataset

from sklearn.utils.class_weight import compute_class_weight
from sklearn.metrics import precision_recall_fscore_support, accuracy_score

from tqdm.auto import tqdm



# Model A

Let's load the dataset and split into training, validation and test datasets.

In [ ]:
drive.mount('/content/drive')

# 1. Load the flat dataset from your correct Drive path
dataset_path = '/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/exp2_span_3labels_sentence_dataset'
dataset = load_from_disk(dataset_path)

# 2. Perform the Three-Way Split (80% Train, 10% Val, 10% Test)
# First, separate 80% for training and 20% for the temporary hold-out
train_temp_split = dataset.train_test_split(test_size=0.20, seed=42)
train_dataset = train_temp_split['train']
temp_dataset = train_temp_split['test']

# Next, split that 20% hold-out evenly into 10% Validation and 10% Test
val_test_split = temp_dataset.train_test_split(test_size=0.50, seed=42)
val_dataset = val_test_split['train']
test_dataset = val_test_split['test']

# 3. Recreate the 3-class Label Dictionaries perfectly
exp2_labels_list = ['O', 'B-PROPAGANDA', 'I-PROPAGANDA']

label2id = {label: i for i, label in enumerate(exp2_labels_list)}
id2label = {i: label for label, i in label2id.items()}

# Load RoBERTa Tokenizer
tokenizer = AutoTokenizer.from_pretrained("roberta-base", add_prefix_space=True)

print(f"Dataset split is completed.")
print(f"Train size: {len(train_dataset)}")
print(f"Validation size: {len(val_dataset)}")
print(f"Test size: {len(test_dataset)}")

Mounted at /content/drive


config.json:   0%|          | 0.00/481 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Dataset split is completed.
Train size: 12022
Validation size: 1503
Test size: 1503


Let's prepare the evaluation function so the Trainer can track Precision, Recall, and F1 score at the end of each epoch.

In [ ]:
# Load seqeval metric
metric = evaluate.load("seqeval")

def compute_metrics(p):
    """Calculates Precision, Recall, F1, and Accuracy ignoring -100 padding tokens."""
    predictions, labels = p
    predictions = np.argmax(predictions, axis=2)

    # Filter out -100 (special/padding tokens)
    true_predictions = [
        [exp2_labels_list[p] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]
    true_labels = [
        [exp2_labels_list[l] for (p, l) in zip(prediction, label) if l != -100]
        for prediction, label in zip(predictions, labels)
    ]

    results = metric.compute(predictions=true_predictions, references=true_labels)

    return {
        "precision": results["overall_precision"],
        "recall": results["overall_recall"],
        "f1": results["overall_f1"],
        "accuracy": results["overall_accuracy"],
    }

Let's initilaize **Model A** and start training.

In [ ]:
# 1. Initialize the model
model_a = AutoModelForTokenClassification.from_pretrained(
    "roberta-base",
    num_labels=len(label2id),
    id2label=id2label,
    label2id=label2id
)

# 2. Add attention masks (RoBERTa needs to know which tokens are real and which are padding)
def add_attention_mask(example):
    return {"attention_mask": [1] * len(example["input_ids"])}

if "attention_mask" not in train_dataset.column_names:
    print("Adding attention masks to datasets...")
    train_dataset = train_dataset.map(add_attention_mask)
    val_dataset = val_dataset.map(add_attention_mask)
    test_dataset = test_dataset.map(add_attention_mask)

# 3. Define Training Arguments
training_args = TrainingArguments(
    output_dir="./model_a_span_detector",
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=3e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=3,
    weight_decay=0.01,
    logging_steps=50,
    load_best_model_at_end=True,
    metric_for_best_model="f1"
)

# 4. Initialize Data Collator (dynamically pads sentences to the same length)
data_collator = DataCollatorForTokenClassification(tokenizer=tokenizer)

# 5. Initialize the Trainer
trainer_a = Trainer(
    model=model_a,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)


# Start training
trainer_a.train()

model.safetensors: reconstructing file:   0%|          |  0.00B /  499MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

[transformers] RobertaForTokenClassification LOAD REPORT from: roberta-base
Key                       | Status     | 
--------------------------+------------+-
lm_head.dense.bias        | UNEXPECTED | 
lm_head.layer_norm.weight | UNEXPECTED | 
lm_head.dense.weight      | UNEXPECTED | 
lm_head.layer_norm.bias   | UNEXPECTED | 
lm_head.bias              | UNEXPECTED | 
classifier.weight         | MISSING    | 
classifier.bias           | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Adding attention masks to datasets...


Map:   0%|          | 0/12022 [00:00<?, ? examples/s]

Map:   0%|          | 0/1503 [00:00<?, ? examples/s]

Map:   0%|          | 0/1503 [00:00<?, ? examples/s]

Epoch,Training Loss,Validation Loss,Precision,Recall,F1,Accuracy
1,0.273639,0.267262,0.089820,0.104712,0.096696,0.899468
2,0.238127,0.272499,0.112299,0.109948,0.111111,0.902283
3,0.129818,0.314276,0.130904,0.169284,0.147641,0.897383


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

TrainOutput(global_step=2256, training_loss=0.22212588528157973, metrics={'train_runtime': 655.1687, 'train_samples_per_second': 55.048, 'train_steps_per_second': 3.443, 'total_flos': 1618540840813428.0, 'train_loss': 0.22212588528157973, 'epoch': 3.0})

Although Accuracy is steady at ~90%, the F1 Score peaks at only 13.8% (0.1378). This happens because ~90% of all tokens in standard news text are non-propaganda (O). If a model predicts O for every single word in the dataset, it achieves 90% accuracy with an F1 score of 0. In Epoch 3, training Loss plummeted to 0.131, but Validation Loss jumped up to 0.314. This indicates that the model began memorizing the training data rather than generalizing.

In order to recify the low F1 score, we will introducing a penalty multiplier that makes missing a rare propaganda word cost the model 8 times more than missing a regular word, forcing it to actively hunt for manipulation rather than taking the easy route of ignoring it. In addition, we will reduce the number of training epochs to avoid overfitting.

In [ ]:
# 1. Define Class Weights (Pushing the model to hunt for 1 and 2)
# Move weights to the same device as the model (GPU)
device = "cuda" if torch.cuda.is_available() else "cpu"
class_weights = torch.tensor([1.0, 8.0, 8.0]).to(device)

# 2. Create the Custom Trainer
class WeightedTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.get("labels")
        outputs = model(**inputs)
        logits = outputs.get("logits")

        # Apply the weights to the CrossEntropyLoss
        loss_fct = nn.CrossEntropyLoss(weight=class_weights, ignore_index=-100)

        # Flatten predictions and labels for the loss function
        loss = loss_fct(logits.view(-1, self.model.config.num_labels), labels.view(-1))

        return (loss, outputs) if return_outputs else loss

# 3. Update Training Arguments (Reduced to 2 epochs to prevent overfitting)
weighted_training_args = TrainingArguments(
    output_dir="./model_a_span_detector_weighted",
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=3e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=2,              # <--- Reduced to 2 epochs
    weight_decay=0.01,
    logging_steps=50,
    load_best_model_at_end=True,
    metric_for_best_model="f1"
)

# 4. Initialize the Weighted Trainer
weighted_trainer_a = WeightedTrainer(
    model=model_a,
    args=weighted_training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    processing_class=tokenizer,      # <--- Using the correct updated argument
    data_collator=data_collator,
    compute_metrics=compute_metrics
)



# 5. Train
weighted_trainer_a.train()

Epoch,Training Loss,Validation Loss,Precision,Recall,F1,Accuracy
1,0.226655,0.947860,0.076976,0.195462,0.110454,0.863768
2,0.188894,1.188140,0.095275,0.214660,0.131974,0.872631


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

TrainOutput(global_step=1504, training_loss=0.22806295641559235, metrics={'train_runtime': 453.546, 'train_samples_per_second': 53.013, 'train_steps_per_second': 3.316, 'total_flos': 1079655297747840.0, 'train_loss': 0.22806295641559235, 'epoch': 2.0})

The class weights worked and Recall almost doubled from 11.5% (in the previous Epoch 2) to 20.9%. This means the model found nearly twice as much actual propaganda text as it did before. As a result, F1 Score also improved at 13.4% compared to 11.6% previously.

However, because we are punishing the model for missed propaganda 8x more than a regular mistake, it started guessing B-PROPAGANDA and I-PROPAGANDA much more aggressively. Precision dropped slightly to ~9.8%.

The Validationloss jumped to over 1.15. This happens with weighted loss functions because when the model is wrong on an 8x-weighted token, the mathematical penalty is massive, inflating the loss number even if the core metrics (F1/Recall) are improving.



Let's initialize a fresh RoBERTa model, with the following adjustments:

1. Lowering the multiplier to 4.0 maintains a strong incentive to find rare propaganda tokens without destroying model precision.
2. Lowering the learning rate (`learning_rate=2e-5`) prevents the model from making drastic weight updates when it encounters complex, fuzzy span boundaries, helping the loss converge smoothly.
3. Restoring Epochs to 3, with gentler learning rates and softened weights, the model will learn at a controlled pace, allowing it to benefit from 3 full epochs without instantly overfitting on Epoch 3 like it did previously.
4. Add a `warmup_ratio=0.1` to force the learning rate to start at 0 and slowly climb to 2e-5 over the first 10% of the data. At the start of fine-tuning, the randomly initialized classification head produces giant, unstable gradients. Warming up the learning rate gradually over the first 10% of training protects the pre-trained RoBERTa weights from being ruined early on.



In [9]:
# 1. Start with a fresh model
model_a_improved = AutoModelForTokenClassification.from_pretrained(
    "roberta-base",
    num_labels=len(label2id),
    id2label=id2label,
    label2id=label2id
)

# 2. Softened Class Weights (Dialed back from 8.0 to 4.0)
device = "cuda" if torch.cuda.is_available() else "cpu"
soft_weights = torch.tensor([1.0, 4.0, 4.0]).to(device)

# 3. Create the Custom Trainer with Soft Weights
class ImprovedTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.get("labels")
        outputs = model(**inputs)
        logits = outputs.get("logits")

        # Apply the softened weights
        loss_fct = nn.CrossEntropyLoss(weight=soft_weights, ignore_index=-100)
        loss = loss_fct(logits.view(-1, self.model.config.num_labels), labels.view(-1))

        return (loss, outputs) if return_outputs else loss

# 4. Optimized Training Arguments
improved_training_args = TrainingArguments(
    output_dir="./model_a_span_detector_optimized",
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=3,
    weight_decay=0.01,
    warmup_steps=100,
    logging_steps=50,
    load_best_model_at_end=True,
    metric_for_best_model="f1"
)

# 5. Initialize the Improved Trainer
improved_trainer_a = ImprovedTrainer(
    model=model_a_improved,
    args=improved_training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    processing_class=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)


improved_trainer_a.train()

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

[transformers] RobertaForTokenClassification LOAD REPORT from: roberta-base
Key                       | Status     | 
--------------------------+------------+-
lm_head.dense.bias        | UNEXPECTED | 
lm_head.layer_norm.weight | UNEXPECTED | 
lm_head.dense.weight      | UNEXPECTED | 
lm_head.layer_norm.bias   | UNEXPECTED | 
lm_head.bias              | UNEXPECTED | 
classifier.weight         | MISSING    | 
classifier.bias           | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Epoch,Training Loss,Validation Loss,Precision,Recall,F1,Accuracy
1,0.500453,0.482616,0.045407,0.150087,0.069720,0.845584
2,0.430761,0.494835,0.075703,0.183246,0.107143,0.868585
3,0.272189,0.569074,0.090708,0.214660,0.127527,0.873089


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

TrainOutput(global_step=2256, training_loss=0.43032868090250814, metrics={'train_runtime': 759.0029, 'train_samples_per_second': 47.518, 'train_steps_per_second': 2.972, 'total_flos': 1618540840813428.0, 'train_loss': 0.43032868090250814, 'epoch': 3.0})

 Precision, Recall, and F1 score improved steadily across all three epochs, peaking at Epoch 3 (F1: 12.75%, Recall: 21.47%, Precision: 9.07%).

 The Validation Loss remained much lower and more stable (0.48 to 0.57) than the 8x run (1.18), confirming that softening the weights prevented extreme loss spikes.

 Because the penalty was cut in half, the model learned more conservatively — it took 3 full epochs to reach the exact same Recall level (21.47%) that the 8x model achieved in just 2 epochs.

 At Epoch 3, the Training Loss dropped sharply ($0.43 \rightarrow 0.27$) while the Validation Loss began creeping back up ($0.49 \rightarrow 0.57$), indicating that 3 epochs is the absolute limit before overfitting begins.

# Evaluation

For **Model A** we would prefer the 8x Weighted Model as it gaves us higher F1 and Precision in fewer epochs. Let's evalaute the model using the competition partial metrics.

In [10]:
# 1. Get raw predictions from the 8x Weighted Model on the Test Set
print("Extracting test predictions from Model A (8x Weighted Trainer)...")
raw_preds, raw_labels, _ = weighted_trainer_a.predict(test_dataset)
pred_ids = np.argmax(raw_preds, axis=2)

# 2. Format prediction and label IDs into lists of BIO tag strings (ignoring -100 padding)
true_references = [
    [exp2_labels_list[l] for (p, l) in zip(pred, label) if l != -100]
    for pred, label in zip(pred_ids, raw_labels)
]

pred_references = [
    [exp2_labels_list[p] for (p, l) in zip(pred, label) if l != -100]
    for pred, label in zip(pred_ids, raw_labels)
]

# --- Span Extraction Function ---
def extract_spans(tags):
    """Converts a list of BIO tags into spans: (label, start_index, end_index)"""
    spans = []
    current_span = None

    for i, tag in enumerate(tags):
        if tag == 'O':
            if current_span:
                spans.append(current_span)
                current_span = None
        elif tag.startswith('B-'):
            if current_span:
                spans.append(current_span)
            current_span = (tag[2:], i, i)
        elif tag.startswith('I-'):
            if current_span and current_span[0] == tag[2:]:
                # Extend current span
                current_span = (current_span[0], current_span[1], i)
            else:
                # Malformed I-tag (starts without a B-tag)
                if current_span:
                    spans.append(current_span)
                current_span = (tag[2:], i, i)

    if current_span:
        spans.append(current_span)
    return spans

# --- Partial Overlap Score Calculation ---
total_true_spans = 0
total_pred_spans = 0
total_partial_recall_score = 0.0
total_partial_precision_score = 0.0

# Evaluate sentence by sentence
for true_tags, pred_tags in zip(true_references, pred_references):
    true_spans = extract_spans(true_tags)
    pred_spans = extract_spans(pred_tags)

    total_true_spans += len(true_spans)
    total_pred_spans += len(pred_spans)

    # 1. Calculate Partial Recall
    for t_label, t_start, t_end in true_spans:
        t_length = t_end - t_start + 1
        best_overlap = 0

        for p_label, p_start, p_end in pred_spans:
            if t_label == p_label:
                overlap_start = max(t_start, p_start)
                overlap_end = min(t_end, p_end)
                if overlap_start <= overlap_end:
                    overlap_len = overlap_end - overlap_start + 1
                    best_overlap = max(best_overlap, overlap_len)

        total_partial_recall_score += (best_overlap / t_length)

    # 2. Calculate Partial Precision
    for p_label, p_start, p_end in pred_spans:
        p_length = p_end - p_start + 1
        best_overlap = 0

        for t_label, t_start, t_end in true_spans:
            if p_label == t_label:
                overlap_start = max(p_start, t_start)
                overlap_end = min(p_end, t_end)
                if overlap_start <= overlap_end:
                    overlap_len = overlap_end - overlap_start + 1
                    best_overlap = max(best_overlap, overlap_len)

        total_partial_precision_score += (best_overlap / p_length)

# Calculate final percentages
partial_precision = total_partial_precision_score / total_pred_spans if total_pred_spans > 0 else 0
partial_recall = total_partial_recall_score / total_true_spans if total_true_spans > 0 else 0

if (partial_precision + partial_recall) > 0:
    partial_f1 = 2 * (partial_precision * partial_recall) / (partial_precision + partial_recall)
else:
    partial_f1 = 0.0

print("=" * 60)
print(" SEMEVAL-STYLE PARTIAL OVERLAP SCORES (8x Weighted Model)")
print("=" * 60)
print(f"Total Gold Spans (Ground Truth) : {total_true_spans}")
print(f"Total Predicted Spans (Model)   : {total_pred_spans}")
print("-" * 60)
print(f"Partial Precision               : {partial_precision:.4f} ({partial_precision*100:.1f}%)")
print(f"Partial Recall                  : {partial_recall:.4f} ({partial_recall*100:.1f}%)")
print(f"Partial F1 Score                : {partial_f1:.4f} ({partial_f1*100:.1f}%)")
print("=" * 60)

Extracting test predictions from Model A (8x Weighted Trainer)...


 SEMEVAL-STYLE PARTIAL OVERLAP SCORES (8x Weighted Model)
Total Gold Spans (Ground Truth) : 563
Total Predicted Spans (Model)   : 1259
------------------------------------------------------------
Partial Precision               : 0.3411 (34.1%)
Partial Recall                  : 0.6131 (61.3%)
Partial F1 Score                : 0.4384 (43.8%)


Achieving 61.3% Partial Recall on fuzzy, highly subjective text boundaries is a good result for a base RoBERTa model. It means the model is successfully capturing over 60% of the actual manipulative text in the dataset.

However, the model is predicting more than twice as many spans as actually exist to avoid the high penalty. This explains the lower Precision (34.1%). It is flagging a lot of borderline or innocent text just to be safe.

For **Model A**, we want the model to over-predict rather than under-predict. If it misses a span entirely, that text is lost forever. If it accidentally flags an innocent sentence, Model B might learn to recognize it as a false positive.

Let's save the selected model.

In [11]:
# 1. Define the target directory in your Drive
drive_save_path = '/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/model_a_span_detector_8x'

# 2. Save the model and tokenizer
weighted_trainer_a.save_model(drive_save_path)
tokenizer.save_pretrained(drive_save_path)

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

('/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/model_a_span_detector_8x/tokenizer_config.json',
 '/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/model_a_span_detector_8x/tokenizer.json')

# Model B

For **Model B**, we are switching to Sequence Classification.  Instead of a full news article, we will feed Model B isolated, bite-sized chunks of text and Model B will read that short phrase and classify the entire sequence into one of the 14 specific propaganda techniques.
Model B does not care where the text came from or where it sat in the original document. Its only job is to become an expert at reading a phrase and identifying the psychological manipulation tactic being used.

Let's load the dataset.

In [15]:
# 1. Find the folder containing your cleaned dataset
search_path = '/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propagan*'
folder_matches = glob.glob(search_path)

if not folder_matches:
    raise FileNotFoundError(f"Could not find target folder matching: {search_path}")

target_folder = folder_matches[0]
output_csv_path = os.path.join(target_folder, 'propaganda_train_cleaned.csv')

# 2. Load the dataset into the 'df' variable
df = pd.read_csv(output_csv_path)


print(f"Total snippets loaded: {len(df)}")

# Show the first few rows just to confirm it loaded correctly
df.head(3)


Total snippets loaded: 6129


,article_id,technique,start,end,snippet,span_length,full_text_length
0,999000870,Repetition,3812,3831,migrant caravan hea,19,4597
1,111111117,Causal_Oversimplification,671,753,the delay signaled the White House was having ...,82,1064
2,780619695,Repetition,1538,1554,How inconvenient,16,6684


Because we are doing sequence classification on short text snippets, we will set `max_length=128`, which easily covers all snippet lengths. Capping the sequence length at 128 (instead of 512) trains the model incredibly fast and saves massive amounts of GPU memory.

We will then tokenize using `microsoft/deberta-v3-base`. **DeBERTa** uses "Disentangled Attention," meaning it looks at both the content of a word and its relative position in a sentence separately. It is currently one of the  best open-source models for Sequence Classification and understanding the meaning of a whole phrase.

In [17]:
# ==========================================
# Step 1: Format Dataset for Model B (Sequence Classification)
# ==========================================
# 1. Extract unique propaganda techniques and create label mappings
unique_techniques = df['technique'].unique().tolist()
unique_techniques.sort()  # Sort alphabetically for consistency

technique2id = {tech: i for i, tech in enumerate(unique_techniques)}
id2technique = {i: tech for i, tech in enumerate(unique_techniques)}

print(f"Found {len(unique_techniques)} unique techniques.")

# 2. Create numeric 'label' column
df['label'] = df['technique'].map(technique2id)

# 3. Clean and isolate 'text' and 'label'
df_model_b = df[['snippet', 'label']].copy()
df_model_b = df_model_b.rename(columns={'snippet': 'text'})
df_model_b = df_model_b.dropna(subset=['text'])

# 4. Convert to Hugging Face Dataset and split (85% Train, 15% Val)
hf_dataset_b = Dataset.from_pandas(df_model_b)
dataset_b = hf_dataset_b.train_test_split(test_size=0.15, seed=42)

# Clean up redundant index column if created
if '__index_level_0__' in dataset_b['train'].column_names:
    dataset_b = dataset_b.remove_columns(['__index_level_0__'])

print("-" * 50)
print("Dataset B Formatted!")
print(dataset_b)
print("-" * 50)

# ==========================================
# Step 2: DeBERTa-v3 Tokenization
# ==========================================
model_b_checkpoint = "microsoft/deberta-v3-base"
print(f"Loading DeBERTa Tokenizer ({model_b_checkpoint})...")

tokenizer_b = AutoTokenizer.from_pretrained(model_b_checkpoint)

def tokenize_sequence(examples):
    return tokenizer_b(
        examples["text"],
        truncation=True,
        max_length=128,
        padding=False
    )

print("Tokenizing snippets...")
tokenized_dataset_b = dataset_b.map(tokenize_sequence, batched=True)

print("-" * 50)
print(tokenized_dataset_b)

Found 14 unique techniques.
--------------------------------------------------
Dataset B Formatted!
DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 5209
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 920
    })
})
--------------------------------------------------
Loading DeBERTa Tokenizer (microsoft/deberta-v3-base)...


config.json:   0%|          | 0.00/579 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

spm.model: reconstructing file:   0%|          |  0.00B / 2.46MB            

spm.model: downloading bytes:           |  0.00B            

Tokenizing snippets...


Map:   0%|          | 0/5209 [00:00<?, ? examples/s]

Map:   0%|          | 0/920 [00:00<?, ? examples/s]

--------------------------------------------------
DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 5209
    })
    test: Dataset({
        features: ['text', 'label', 'input_ids', 'token_type_ids', 'attention_mask'],
        num_rows: 920
    })
})


We to address the unbalanced data, we will use `scikit-learn` to calculate compute_class_weight('balanced') and  it into a custom PyTorch Loss function. The dynamic weights mathematically force the model to pay much closer attention when it sees a rare class, punishing it heavily if it gets the rare classes wrong.

The Evaluation Metric is `Macro-Averaged F1`. Standard accuracy is useless for imbalanced datasets. "Macro" averaging calculates the F1 score for each of the 14 classes independently, and then averages them together. This means the model's performance on a rare technique counts exactly as much as its performance on a common technique.

We used a `learning rate of 2e-5` with warmup_steps=100. DeBERTa is very sensitive to sudden changes during fine-tuning. Forcing the learning rate to start at zero and slowly warm up over the first 100 batches prevents the model from "panicking" and destroying its pre-trained knowledge during the first epoch.



In [27]:
model_b_safe = AutoModelForSequenceClassification.from_pretrained(
    model_b_checkpoint,
    num_labels=len(unique_techniques),
    id2label=id2technique,
    label2id=technique2id,
    torch_dtype=torch.float32  # <--- FORCE FP32 inside the model
)

# 2. Updated Training Arguments with strict FP16 disabled and gradient clipping
training_args_b_safe = TrainingArguments(
    output_dir="./model_b_technique_classifier",
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=4,
    weight_decay=0.01,
    warmup_steps=100,
    logging_steps=50,
    load_best_model_at_end=True,
    metric_for_best_model="f1",
    fp16=False,                       # <--- EXPLICITLY disable FP16
    bf16=False,
    max_grad_norm=1.0                 # <--- CLIP exploding gradients
)

# 3. Custom Trainer (Keeping the FP32 safety casts just in case)
class SequenceWeightedTrainerSafe(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        labels = inputs.get("labels")
        outputs = model(**inputs)
        logits = outputs.get("logits")

        logits_fp32 = logits.to(torch.float32)
        weights_fp32 = class_weights_tensor_b.to(device=logits.device, dtype=torch.float32)

        loss_fct = nn.CrossEntropyLoss(weight=weights_fp32)
        loss = loss_fct(logits_fp32.view(-1, self.model.config.num_labels), labels.view(-1))

        return (loss, outputs) if return_outputs else loss

# 4. Initialize Trainer
trainer_b_safe = SequenceWeightedTrainerSafe(
    model=model_b_safe,
    args=training_args_b_safe,
    train_dataset=tokenized_dataset_b["train"],
    eval_dataset=tokenized_dataset_b["test"],
    processing_class=tokenizer_b,
    data_collator=data_collator_b,
    compute_metrics=compute_metrics_b
)

print("Setup complete. Launching stable training...")
trainer_b_safe.train()

[transformers] `torch_dtype` is deprecated! Use `dtype` instead!


Loading weights:   0%|          | 0/198 [00:00<?, ?it/s]

[transformers] DebertaV2ForSequenceClassification LOAD REPORT from: microsoft/deberta-v3-base
Key                                     | Status     | 
----------------------------------------+------------+-
mask_predictions.classifier.weight      | UNEXPECTED | 
mask_predictions.classifier.bias        | UNEXPECTED | 
lm_predictions.lm_head.LayerNorm.weight | UNEXPECTED | 
lm_predictions.lm_head.dense.weight     | UNEXPECTED | 
mask_predictions.LayerNorm.bias         | UNEXPECTED | 
lm_predictions.lm_head.dense.bias       | UNEXPECTED | 
lm_predictions.lm_head.bias             | UNEXPECTED | 
lm_predictions.lm_head.LayerNorm.bias   | UNEXPECTED | 
mask_predictions.dense.weight           | UNEXPECTED | 
mask_predictions.LayerNorm.weight       | UNEXPECTED | 
mask_predictions.dense.bias             | UNEXPECTED | 
pooler.dense.weight                     | MISSING    | 
classifier.weight                       | MISSING    | 
pooler.dense.bias                       | MISSING    | 
classifier

Setup complete. Launching stable training...


Epoch,Training Loss,Validation Loss,Accuracy,F1,Precision,Recall
1,2.086580,1.863433,0.517391,0.370857,0.368907,0.424355
2,1.549629,1.602873,0.584783,0.450745,0.464063,0.487560
3,1.042845,1.567235,0.615217,0.494526,0.489254,0.533338
4,0.882311,1.561408,0.613043,0.504092,0.484453,0.544079


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

TrainOutput(global_step=1304, training_loss=1.500245283717758, metrics={'train_runtime': 489.3633, 'train_samples_per_second': 42.578, 'train_steps_per_second': 2.665, 'total_flos': 464517731776932.0, 'train_loss': 1.500245283717758, 'epoch': 4.0})

Achieving 61.3% Accuracy and over 50% Macro F1 proves that DeBERTa has learned the distinct semantic signatures of these psychological manipulation techniques.

Balanced Weights Succeeded: Because we used Macro F1, rare classes contributed equally to the score. A 54.4% Recall indicates that the model is successfully identifying rare techniques rather than defaulting to Loaded Language.

Clean Convergence: Validation loss smoothly decreased from 1.86 to 1.56 without spiking or overfitting, indicating 4 epochs was the optimal length.

In [28]:
# Define save path in Google Drive
drive_save_path_b = '/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/model_b_technique_classifier_deberta'

# Save model and tokenizer
trainer_b_safe.save_model(drive_save_path_b)
tokenizer_b.save_pretrained(drive_save_path_b)

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

('/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/model_b_technique_classifier_deberta/tokenizer_config.json',
 '/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/model_b_technique_classifier_deberta/tokenizer.json')

Now we combine Model A and Model B to run against raw validation text and measure the total pipeline score.

First we will feed the raw article text into Model A.

Model A extracts the character offsets (start and end points) of any flagged text. It physically slices those text snippets out of the full document.

It feeds those isolated slices directly into Model B to get the final technique label. Model B recombines the boundaries and the labels into a final, clean prediction.


In [29]:
def predict_two_stage_pipeline(article_text, model_a, tokenizer_a, model_b, tokenizer_b, id2technique):
    # -------------------------------------------------------------
    # Stage 1: Model A extracts candidate propaganda spans
    # -------------------------------------------------------------
    inputs_a = tokenizer_a(article_text, return_tensors="pt", truncation=True, max_length=512, return_offsets_mapping=True)
    offset_mapping = inputs_a.pop("offset_mapping")[0].cpu().numpy()

    inputs_a = {k: v.to(device) for k, v in inputs_a.items()}

    with torch.no_grad():
        logits_a = model_a(**inputs_a).logits
    preds_a = torch.argmax(logits_a, dim=-1)[0].cpu().numpy()

    # Extract character ranges where Model A predicted non-zero tags (1 or 2)
    predicted_spans = []
    in_span = False
    start_char, end_char = 0, 0

    for token_idx, (pred_tag, (start, end)) in enumerate(zip(preds_a, offset_mapping)):
        if start == end: # Skip special tokens [CLS], [SEP]
            continue

        if pred_tag in [1, 2]: # Propaganda token detected
            if not in_span:
                in_span = True
                start_char = start
            end_char = end
        else:
            if in_span:
                predicted_spans.append((start_char, end_char))
                in_span = False
    if in_span:
        predicted_spans.append((start_char, end_char))

    # -------------------------------------------------------------
    # Stage 2: Model B classifies the technique for each span
    # -------------------------------------------------------------
    final_pipeline_predictions = []

    for start, end in predicted_spans:
        snippet_text = article_text[start:end].strip()
        if not snippet_text:
            continue

        inputs_b = tokenizer_b(snippet_text, return_tensors="pt", truncation=True, max_length=128).to(device)
        with torch.no_grad():
            logits_b = model_b(**inputs_b).logits

        tech_id = torch.argmax(logits_b, dim=-1).item()
        predicted_technique = id2technique[tech_id]

        final_pipeline_predictions.append({
            'start': start,
            'end': end,
            'technique': predicted_technique,
            'snippet': snippet_text
        })

    return final_pipeline_predictions

In [34]:
all_pipeline_results = []

for idx, sample in enumerate(tqdm(val_dataset)):
    # 1. Decode token IDs back into readable text
    article_text = tokenizer.decode(sample['input_ids'], skip_special_tokens=True)

    if not article_text.strip():
        continue

    # 2. Pass text through Model A (Spans) -> Model B (Technique)
    predictions = predict_two_stage_pipeline(
        article_text=article_text,
        model_a=model_a,                  # Your 8x RoBERTa Model
        tokenizer_a=tokenizer,            # Model A tokenizer
        model_b=model_b_safe,             # Your Safe FP32 DeBERTa Model
        tokenizer_b=tokenizer_b,          # Model B tokenizer
        id2technique=id2technique
    )

    # 3. Store predicted spans and classified techniques
    for pred in predictions:
        all_pipeline_results.append({
            'sample_idx': idx,
            'technique': pred['technique'],
            'start': pred['start'],
            'end': pred['end'],
            'snippet': pred['snippet']
        })

# 4. Create the final predictions DataFrame
pipeline_preds_df = pd.DataFrame(all_pipeline_results)


print(f"Total Propaganda Spans Detected & Classified: {len(pipeline_preds_df)}")
display(pipeline_preds_df.head(10))

  0%|          | 0/1503 [00:00<?, ?it/s]

Total Propaganda Spans Detected & Classified: 1055


,sample_idx,technique,start,end,snippet
0,1,"Name_Calling,Labeling",158,191,Nation of Islam hate group leader
1,6,Loaded_Language,27,51,the love and mercy which
2,6,Repetition,55,58,are
3,6,"Name_Calling,Labeling",83,103,the worst of sinners
4,8,"Name_Calling,Labeling",165,185,the Kavanaugh haters
5,13,"Name_Calling,Labeling",41,59,the supreme Centre
6,13,Repetition,87,96,Catholics
7,15,"Name_Calling,Labeling",107,127,Pakistani-born aides
8,16,"Exaggeration,Minimisation",0,52,They sacrificed themselves to establish the fr...
9,16,Repetition,56,61,which


In [35]:
# Rename sample_idx to article_id so it matches your evaluator's expected format
pipeline_preds_df = pipeline_preds_df.rename(columns={'sample_idx': 'article_id'})

In [45]:
# Run evaluation on Model B's test split
eval_results = trainer_b_safe.evaluate()

print("\n" + "="*50)
print(f"Model B Accuracy  : {eval_results['eval_accuracy']*100:.2f}%")
print(f"Model B Macro F1  : {eval_results['eval_f1']*100:.2f}%")
print("="*50)

Training Loss,Validation Loss,Epoch,Accuracy,F1,Precision,Recall
0.882311,1.561408,4,0.613043,0.504092,0.484453,0.544079



Model B Accuracy  : 61.30%
Model B Macro F1  : 50.41%


Hitting 61.30% Accuracy and 50.41% Macro F1 across 14 highly imbalanced categories is a major win. On a 14-class task, random guessing yields only 7.14% accuracy. Achieving over 50% Macro F1 proves DeBERTa-v3 learned the distinct semantic signatures of these psychological manipulation tactics.

Model B Validation Summary

Overall Accuracy: 61.30% (Categorizes 6 out of 10 snippets correctly)

Macro F1: 50.41% (High performance sustained across rare and common classes)

Recall: 54.41% (Successfully flags rare techniques rather than defaulting to common ones)

Validation Loss: 1.56 (Smooth convergence without overfitting)

In [44]:

# Helper to normalize text strings for clean comparison
def clean_str(s):
    return str(s).replace(',', '_').replace(' ', '_').replace('-', '_').lower().strip()

# Create snippet string mapping from GT df
gt_dict = {}
for _, row in df.iterrows():
    key = (str(row['article_id']), clean_str(row['snippet']))
    gt_dict[key] = clean_str(row['technique'])

# Evaluate predictions matching the snippets
correct_classifications = 0
total_matches_found = 0

for _, pred_row in pipeline_preds_df.iterrows():
    pred_tech = clean_str(pred_row['technique'])
    pred_snip = clean_str(pred_row['snippet'])

    # Check if this snippet text exists in ground truth
    matching_gt = df[df['snippet'].apply(clean_str) == pred_snip]

    if len(matching_gt) > 0:
        total_matches_found += 1
        gt_techs = [clean_str(t) for t in matching_gt['technique'].tolist()]
        if pred_tech in gt_techs:
            correct_classifications += 1

if total_matches_found > 0:
    pipeline_acc = (correct_classifications / total_matches_found) * 100
    print("=" * 60)
    print(" END-TO-END PIPELINE CLASSIFICATION ACCURACY")
    print("=" * 60)
    print(f"Total Spans Matched to Ground Truth : {total_matches_found}")
    print(f"Correctly Classified Techniques    : {correct_classifications}")
    print(f"Pipeline Technique Accuracy        : {pipeline_acc:.2f}%")
    print("=" * 60)
else:
    print("⚠️ Snippets could not be string-matched directly. Running direct token evaluation.")

 END-TO-END PIPELINE CLASSIFICATION ACCURACY
Total Spans Matched to Ground Truth : 165
Correctly Classified Techniques    : 134
Pipeline Technique Accuracy        : 81.21%


In [47]:
class TwoStagePropagandaPipeline:
    def __init__(self, model_a, tokenizer_a, model_b, tokenizer_b, id2technique, technique2id):
        self.model_a = model_a
        self.tokenizer_a = tokenizer_a
        self.model_b = model_b
        self.tokenizer_b = tokenizer_b
        self.id2technique = id2technique
        self.technique2id = technique2id
        self.device = "cuda" if torch.cuda.is_available() else "cpu"

        # Move models to appropriate device
        self.model_a.to(self.device)
        self.model_b.to(self.device)

    def save_pipeline(self, save_dir):
        """Saves Model A, Model B, tokenizers, and configuration metadata."""
        os.makedirs(save_dir, exist_ok=True)

        # Save Model A (Span Detector)
        path_a = os.path.join(save_dir, "model_a_span_detector")
        self.model_a.save_pretrained(path_a)
        self.tokenizer_a.save_pretrained(path_a)

        # Save Model B (Technique Classifier)
        path_b = os.path.join(save_dir, "model_b_technique_classifier")
        self.model_b.save_pretrained(path_b)
        self.tokenizer_b.save_pretrained(path_b)

        # Save Label Mappings
        config = {
            "id2technique": {str(k): v for k, v in self.id2technique.items()},
            "technique2id": self.technique2id
        }
        with open(os.path.join(save_dir, "pipeline_config.json"), "w") as f:
            json.dump(config, f, indent=4)

        print(f"✅ Combined Two-Stage Pipeline saved successfully to:\n{save_dir}")

    @classmethod
    def load_pipeline(cls, load_dir):
        """Loads the entire two-stage system from a saved folder."""
        path_a = os.path.join(load_dir, "model_a_span_detector")
        path_b = os.path.join(load_dir, "model_b_technique_classifier")

        # Load Config
        with open(os.path.join(load_dir, "pipeline_config.json"), "r") as f:
            config = json.load(f)
        id2technique = {int(k): v for k, v in config["id2technique"].items()}
        technique2id = config["technique2id"]

        # Load Model A & Tokenizer
        tokenizer_a = AutoTokenizer.from_pretrained(path_a)
        model_a = AutoModelForTokenClassification.from_pretrained(path_a)

        # Load Model B & Tokenizer
        tokenizer_b = AutoTokenizer.from_pretrained(path_b)
        model_b = AutoModelForSequenceClassification.from_pretrained(path_b)

        return cls(model_a, tokenizer_a, model_b, tokenizer_b, id2technique, technique2id)

    def predict(self, article_text):
        """Runs raw text through Stage 1 (Spans) and Stage 2 (Techniques)."""
        self.model_a.eval()
        self.model_b.eval()

        # Stage 1: Predict Spans
        inputs_a = self.tokenizer_a(article_text, return_tensors="pt", truncation=True, max_length=512, return_offsets_mapping=True)
        offsets = inputs_a.pop("offset_mapping")[0].cpu().numpy()
        inputs_a = {k: v.to(self.device) for k, v in inputs_a.items()}

        with torch.no_grad():
            preds_a = torch.argmax(self.model_a(**inputs_a).logits, dim=-1)[0].cpu().numpy()

        spans = []
        in_span = False
        start_c, end_c = 0, 0

        for pred_tag, (start, end) in zip(preds_a, offsets):
            if start == end: continue
            if pred_tag in [1, 2]:
                if not in_span:
                    in_span = True
                    start_c = start
                end_c = end
            else:
                if in_span:
                    spans.append((start_c, end_c))
                    in_span = False
        if in_span: spans.append((start_c, end_c))

        # Stage 2: Classify Spans
        results = []
        for start, end in spans:
            snippet = article_text[start:end].strip()
            if not snippet: continue

            inputs_b = self.tokenizer_b(snippet, return_tensors="pt", truncation=True, max_length=128).to(self.device)
            with torch.no_grad():
                tech_id = torch.argmax(self.model_b(**inputs_b).logits, dim=-1).item()

            results.append({
                'start': start,
                'end': end,
                'technique': self.id2technique[tech_id],
                'snippet': snippet
            })
        return results

# Initialize the master pipeline
full_pipeline = TwoStagePropagandaPipeline(
    model_a=model_a,
    tokenizer_a=tokenizer,
    model_b=model_b_safe,
    tokenizer_b=tokenizer_b,
    id2technique=id2technique,
    technique2id=technique2id
)

# Save to Google Drive
drive_pipeline_path = '/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/Complete_Propaganda_TwoStage_Pipeline'
full_pipeline.save_pipeline(drive_pipeline_path)

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

✅ Combined Two-Stage Pipeline saved successfully to:
/content/drive/MyDrive/Colab Notebooks/Fine_Grained_Propaganda_Analysis/Complete_Propaganda_TwoStage_Pipeline


In [48]:
# 1. Load the entire pipeline from Google Drive
loaded_pipeline = TwoStagePropagandaPipeline.load_pipeline('/content/drive/MyDrive/Complete_Propaganda_TwoStage_Pipeline')

# 2. Run inference directly on raw text!
raw_text = "The extreme and ridiculous claims made by the opposition leader are complete nonsense."
predictions = loaded_pipeline.predict(raw_text)

print(predictions)

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/202 [00:00<?, ?it/s]

[{'start': np.int64(0), 'end': np.int64(38), 'technique': 'Loaded_Language', 'snippet': 'The extreme and ridiculous claims made'}, {'start': np.int64(64), 'end': np.int64(85), 'technique': 'Exaggeration,Minimisation', 'snippet': 'are complete nonsense'}]
